### ISSUE TITLE
"Xs must only contain ascending values" thrown at random.

### ISSUE THREAD

**Issue by fkelava** [ScottPlot/ScottPlot#906]
"Xs must only contain ascending values" will sometimes throw for no discernible reason, even when all Xs _are_ strictly ascending values. I cannot determine any realistic cause for this error to be thrown.

This happens in ScottPlot 4.1.9 and 4.1.10-beta.

Reproduce by doing the following:

1. Create a SignalXY plot.
2. Have Xs be filled by a constantly increasing integer, such that no two values are the same.
3. Have Ys be whatever.
4. Plot it. In my instance, I keep receiving "Xs must only contain ascending values"- except it already _does!_

The image illustrates a breakpoint reached upon the exception triggering and showcases a portion of the Xs array. 
A snippet is provided below.

![image](https://user-images.githubusercontent.com/28020919/112736978-245e0400-8f57-11eb-86ee-ba1c191e4b9f.png)

For context in the snippet- the timestamp of a ``CanSignal`` is normally the current Unix timestamp in milliseconds. No two messages may share one, so uniqueness is guaranteed. However, both that approach and a test long that is constantly increased as messages arrive (_and was used in the image_) yield the same error.

I'm at a loss. I'm certain ScottPlot hasn't magically stopped working, but the nature of the error eludes me. Please sanity-check my stuff.

```cs
private readonly SignalPlotXY _sigPlot;
private readonly List<double> _valueBuf;
private readonly List<double> _timestampBuf;
private readonly int _valueLimit;

private object _plotRenderLock;

private void PlotData(object sender, PlotReadyEventArgs e) {
  lock (_plotRenderLock) {
    foreach (CanSignal sig in e.Buffer) {
      _valueBuf.Add(sig.Value);
      _timestampBuf.Add(sig.Timestamp);
    }

    _sigPlot.Xs = _timestampBuf.ToArray();
    _sigPlot.Ys = _valueBuf.ToArray();
    _sigPlot.MaxRenderIndex = _sigPlot.Ys.Length - 1;

    Graph.Render();
    Graph.Refresh();
    Graph.Plot.AxisAuto(0.02D, 0.02D);
                
    if (_valueBuf.Count >= _valueLimit) {
      _valueBuf.RemoveRange(0, e.Buffer.Count);
      _timestampBuf.RemoveRange(0, e.Buffer.Count);
    }
  }
}

```

**Comment by bclehmann** [comment]
I don't know what platform you're on or what's sending the events, but my first guess is that the events are being received out of order. This is especially likely if the events can come from a different thread.

One easy way of checking if this is the issue and not ScottPlot is to create a short for loop that tests this condition. Something like this:

```cs
for(int i = 1; i < xs.Length; i++){
    if(xs[i] < xs[i - 1]){
        throw new ArgumentException($"Values: {xs[i - 1], {xs[i]}}");
    }
}
```

You can then see if your input is indeed in ascending order.

Note you'll have to have this loop run _before_ you set `_sigPlot.Xs`, as the input validation runs on the assignment, not on render.

Also, depending on how many points you have, a `ScatterPlot` may make sense. It has worse performance on large datasets, but it doesn't require the x values be ascending.

**Comment by StendProg** [comment]
Hi @fkelava , @bclehmann ,

In the example you provided, something is wrong with the item at index 256.
And you should look at the `value[]` in the debugger, instead of `_Xs`. Assignment goes at the very end if all checks are passed.

As for the error itself, I think it is 99% in your data, the verification itself is very simple and can hardly contain an error.

**Comment by fkelava** [comment]
@bclehmann, @StendProg Thank you- both of your answers were extremely helpful.

I ran @bclehmann's check and it threw no exception. Indeed, the data I was feeding it through the event _was_ in ascending order, but for some reason the first value and timestamp in the buffer reappeared at the final index (256 in this instance). I'm uncertain as to why that is, but I'll get to the bottom of it and patch it up.

> I'm at a loss. I'm certain ScottPlot hasn't magically stopped working, but the nature of the error eludes me. Please sanity-check my stuff.

It would seem this premonition was correct. I'm still incredibly confused as to how this has come to pass, but at least now I have verified the source of the issue. Thank you for your help.
